python/mxnet/optimizer.py

```diff
@@ -171,6 +172,130 @@ def update(self, index, weight, grad, state):
 
 @register
+class Adam(Optimizer):
+    """Adam optimizer as described in [King2014]_.
+
+    .. [King2014] Diederik Kingma, Jimmy Ba,
+       *Adam: A Method for Stochastic Optimization*,
+       http://arxiv.org/abs/1412.6980
+
+    the code in this class was adapted from
+    https://github.com/mila-udem/blocks/blob/master/blocks/algorithms/__init__.py#L765
+
+    Parameters
+    ----------
+    learning_rate : float, optional
+        Step size.
+        Default value is set to 0.002.
+    beta1 : float, optional
+        Exponential decay rate for the first moment estimates.
+        Default value is set to 0.9.
+    beta2 : float, optional
+        Exponential decay rate for the second moment estimates.
+        Default value is set to 0.999.
+    epsilon : float, optional
+        Default value is set to 1e-8.
+    decay_factor : float, optional
+        Default value is set to 1 - 1e-8.
+
+    wd : float, optional
+        L2 regularization coefficient add to all the weights
+    rescale_grad : float, optional
+        rescaling factor of gradient.
+
+    clip_gradient : float, optional
+        clip gradient in range [-clip_gradient, clip_gradient]
+    """
+    def __init__(self, learning_rate=0.002,
+                 beta1=0.9, beta2=0.999, epsilon=1e-8,
+                 decay_factor=(1 - 1e-8),
+                 wd=0.,
+                 rescale_grad=1, clip_gradient=None,
+                 lr_scheduler=None):
+        super(Adam, self).__init__(rescale_grad)
+        self.lr = learning_rate
+        self.beta1 = beta1
+        self.beta2 = beta2
+        self.epsilon = epsilon
+        self.decay_factor = decay_factor
+        self.wd = wd
+        self.clip_gradient = clip_gradient
+        self.lr_scheduler = lr_scheduler
+        if lr_scheduler is not None:
+            self.lr_scheduler.base_lr = learning_rate
+        self.time = 0
+        self.time_first_index = None
+
+    def create_state(self, index, weight):
+        """Create additional optimizer state: mean, variance
+
+        Parameters
+        ----------
+        weight : NDArray
+            The weight data
+
+        """
+        self.time_first_index = None  # time is incremented only on the first index
+        return (zeros(weight.shape, weight.context),  # mean
+                zeros(weight.shape, weight.context))  # variance
+
+    def update(self, index, weight, grad, state):
+        """Update the parameters.
+
+        Parameters
+        ----------
+        index : int
+            An unique integer key used to index the parameters
+
+        weight : NDArray
+            weight ndarray
+
+        grad : NDArray
+            grad ndarray
+
+        state : NDArray or other objects returned by init_state
+            The auxiliary state used in optimization.
+        """
+        assert(isinstance(weight, NDArray))
+        assert(isinstance(grad, NDArray))
+        if self.lr_scheduler is not None:
+            lr = self.lr_scheduler(self.epoch)
```

Comment:
(looks like i lost my last comment)

could you please first merge the current master, and then replace the above line by

```python
            lr = self.lr_scheduler(self.num_update)
            self._update_count(index)
```
refer to https://github.com/dmlc/mxnet/blob/master/python/mxnet/optimizer.py#L158